rfcs/01391_async_sdk.md

```diff
@@ -0,0 +1,189 @@
+    // channel is automatically closed.
+}
+```
+
+## Executor and primitive internals
+When a `ChannelRead` is polled and no data is readily available, it must request the executor to wake it once data does become available. We take a similar approach to other executors like [`tokio`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/context.rs#L7) here, and make the executor available to `ChannelRead` via thread-local storage (TLS). Note that for Webassembly this is equivalent to a global variable, since threads are not standardized or widely supported. While the initial version of the executor will be single-threaded, the API is designed to allow for multithreaded execution.
+
+The executor maintains a set of [`Waker`](https://doc.rust-lang.org/std/task/struct.Waker.html) objects and the channels associated with them, which the `ChannelRead` futures add themselves to when polled. The executor runs the following loop:
+```rust
+loop {
+    // futures_pool: Set of the main event loop future and all `spawn`ed futures.
+    // channels: Set of all channels that have wakers associated with them. Populated by `ChannelRead`.
+    // wakers: Set of all wakers. Populated by `ChannelRead`.
+
+    // Returns when no future can make progress
+    futures_pool.run_until_stalled();
+
+    if futures_pool.is_empty() {
+        break;
+    }
+
+    let channel_read_statuses = wait_on_channels(&channels);
+
+    // For each ready channel, awake one of the futures that was waiting on it.
+    wake_if_ready(channel_read_statuses, wakers);
+}
+``` 
+
+Implementing an executor for scratch is hard, unsafe and error-prone, even for a [very simple basic](https://github.com/spacejam/extreme#bugs-encountered-over-time). Instead, we can use an existing and tested [futures pool](https://docs.rs/futures/0.3.5/futures/executor/struct.LocalPool.html) that is part of the rust-lang supported `futures` crate, and implement our executor on top of that. This greatly reduces engineering effort, unsafe code and testing surface.
+
+# Planning
+The initial implementation work for the Async SDK will be done in a separate `oak_async` crate is under `experimental/`. Most of the changes are self-contained and can easily live in a separate crate. An extension trait `ReceiverAsyncExt` can add the required methods to `oak::io::Receiver`. The `hello_world` and `chat` examples will be used to demonstrate and test the new functionality. 
+
+Rust's futures are a zero-cost abstraction, and even in an `async` function it is possible to block and call `wait_on_channels` manually (though not recommended if asynchronous execution is the goal). This means we can make `async` the default, while still allowing users to call blocking APIs if they insist on doing so. Convenient methods for blocking such as [`Receiver::receive`](https://project-oak.github.io/oak/sdk/doc/oak/io/trait.ReceiverExt.html#tymethod.receive) will be removed from the SDK to prevent users from accidentally blocking.
```

Comment:
> will be removed from the SDK

Maybe we should have a couple of feature gates (`"sync"`, `"async"`) in the `oak` SDK crate - wdyt?

Reply:
I'm a bit unsure what the benefit would be of having such feature gates. If sync code can run just fine in an async block, why keep a separate sync feature around?